### PR TITLE
[SPARK-20301][FLAKY-TEST] Fix Hadoop Shell.runCommand flakiness in Structured Streaming tests

### PR DESCRIPTION
**Pull request by brkyvz**
## What changes were proposed in this pull request?

Some Structured Streaming tests show flakiness such as:
```
[info] - prune results by current_date, complete mode - 696 *** FAILED *** (10 seconds, 937 milliseconds)
[info]   Timed out while stopping and waiting for microbatchthread to terminate.: The code passed to failAfter did not complete within 10 seconds.
```

This happens when we wait for the stream to stop, but it doesn't. The reason it doesn't stop is that we interrupt the microBatchThread, but Hadoop's `Shell.runCommand` swallows the interrupt exception, and the exception is not propagated upstream to the microBatchThread. Then this thread continues to run, only to start blocking on the `streamManualClock`.

## How was this patch tested?

Thousand retries locally and [Jenkins](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75720/testReport) of the flaky tests

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 1]
```diff
@@ -284,42 +284,38 @@ class StreamExecution(
...
-                updateStatusMessage("Waiting for data to arrive")
-                Thread.sleep(pollingDelayMs)
+                currentStatus = currentStatus.copy(isDataAvailable = true)
+                updateStatusMessage("Processing new data")
+                runBatch(sparkSessionToRunBatches)
               }
-              true
+            }
+            // Report trigger as finished and construct progress object.
+            finishTrigger(dataAvailable)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: why did you move this out of the `reportTimeTaken { ... }`?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamTest.scala [hunk 2]
```diff
@@ -277,6 +277,11 @@ trait StreamTest extends QueryTest with SharedSQLContext with Timeouts {
 
     def threadState =
       if (currentStream != null && currentStream.microBatchThread.isAlive) "alive" else "dead"
+    def threadStackTrace = if (currentStream != null && currentStream.microBatchThread.isAlive) {
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamTest.scala]: +1 on keeping this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala (hunk 1 above)
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: I don't think I moved it out. Is the diff and whitespace confusing?

**Comment by brkyvz** [comment]
retest this please

**Comment by tdas** [comment]
LGTM. Merging to master
